src/blog/understanding-ecmascript-part-4.md

```diff
@@ -0,0 +1,280 @@
+
+```grammar
+AssignmentExpression :
+...
+ConditionalExpression (eventually leading to PrimaryExpression)
+ArrowFunction
+
+PrimaryExpression :
+...
+ParenthesizedExpression
+
+ArrowFunction :
+ArrowParameterList => ConciseBody
+```
+
+We couldn't choose the correct production with limited lookahead. Imagine we had to parse a `AssignmentExpression` and the next token is `(`. How would we decide what to parse next? We could either parse an `ParenthesizedExpression` or an `ArrowParameterList`, but our guess could go wrong.
+
+### The very permissive new symbol: CPEAAPL
+
+We'd like to specify the grammar in such a way that it's possible to parse JavaScript according to it with limited lookahead.
+
+The spec solves this problem by introducing the symbol `CoverParenthesizedExpressionAndArrowParameterList` (`CPEAAPL` for short). `CPEAAPL` is a symbol that is actually an `ParenthesizedExpression` or an `ArrowParameterList` behind the scenes, but we don't yet know which one.
+
+The [productions](https://tc39.es/ecma262/#prod-CoverParenthesizedExpressionAndArrowParameterList) for `CPEAAPL` are very permissive, allowing all constructs that can occur in `ParenthesizedExpression`s and in `ArrowParameterList`s:
+
+```grammar
+CPEAAPL :
+( Expression )
+( Expression , )
+( )
+(... BindingIdentifier )
+(... BindingPattern )
+( Expression , ... BindingIdentifier )
+( Expression , ... BindingPattern )
+```
+
+For example, the following expressions are valid `CPEAAPL`s:
+
+```javascript
+// Valid ParenthesizedExpression and ArrowParameterList:
+(a, b)
+(a, b = 1)
+
+// Valid ParenthesizedExpression:
+(1, 2, 3)
+(function foo() { })
+
+// Valid ArrowParameterList:
+()
+(a, b,)
+(a, ...b)
+(a = 1, ...b)
+
+// Not valid either, but still a CPEAAPL:
+(1, ...b)
+(1, )
+```
+
+Trailing comma and the `...` can occur only in `ArrowParameterList`. Some constructs, like `b = 1` can occur in both, but they have different meanings: Inside `ParenthesizedExpression` it's an assignment, inside `ArrowParameterList` it's a parameter with a default value.
+
+### Using CPEAAPL in grammar rules
+
+Now we can use the very permissive `CPEAAPL` in grammar productions:
+
+```grammar
+AssignmentExpression :
+ConditionalExpression (eventually leading to PrimaryExpression)
+ArrowFunction
+...
+
+ArrowFunction :
+ArrowParameters => ConciseBody
+
+ArrowParameters :
+BindingIdentifier
+CPEAAPL
+
+PrimaryExpression :
+...
+CPEAAPL
+
+```
+
+Imagine we're again in the situation that we need to parse an `AssignmentExpression` and the next token is `(`. Now we can just decide to parse a `CPEAAPL` and figure out later what it actually is. It doesn't matter whether we're parsing an `ArrowFunction` or a `ParenthesizedExpression`, the next symbol to parse is `CPEAAPL` in any case!
+
+After we've parsed the `CPEAAPL`, we can decide whether the original `AssignmentExpression` is an `ArrowFunction` or a `ParenthesizedExpression` based on the token following the `CPEAAPL`.
```

Comment:
Pedantic aside: this isn't quite what the "covers" semantics in ecma262 does.

It doesn't parse CPEAAPL and then disambiguates based on the following token. It parses the entire program into an AST first. So, the AST is already something like (ArrowFunction (ArrowParameters (CPEAAPL ...)) `=>` (ConciseBody ...)). Then it attempts to refine the token list that compromises CPEAAPL into ArrowFormalParameters. If this "covering" of ArrowFormalParameters succeeds on that token list without any tokens left over, then we have a successful parse. 

In practice, the easiest way to understand this is to disambiguate using the immediately following non-terminal. But, it'd be good to call out this distinction, since the next section is all about restrictions. None of the Supplemental Syntax sections and covering restrictions talk about non-terminals outside of the production at all, so the reader might wonder "wait, I thought you said these were disambiguated by looking at what follows".

Reply:
I never said it was :) 

I''ll need to clarify this, I'll write my thoughts here.

There are two things:

1) the existence of CPEAAPL enables us to go ahead with parsing and pick the production to follow after having parsed the CPEAAPL

2) after we have the AST (having the AST requires having decided whether we have an ArrowFunction or ParenthesizedExpression), we apply the "covering" rules like you said.

-> How to makes this clearer in the text?

2 is the "cover semantics" but 1 is a needed step for constructing the AST in the first place.

Reply:
Ah, right, I might see what you mean. This sentence is confusing: "Once we know whether we were parsing an `ArrowFunction` or `ParenthesizedExpression`, we need to disallow the corresponding illegal construct"

-> From this one might think that this is done right away, after disambiguating... let's see how I can fix this.

Reply:
I tried to clarify this, by making the section "Using CPEAAPL" only talk about how to disambiguate during parsing, and the section "Restricting CPEAAPLs" only talk about the covering. PTAL

Reply:
Hmm, I still think "we can decide whether the original `AssignmentExpression` is an `ArrowFunction` or a `ParenthesizedExpression` based on the token following the `CPEAAPL`," is misleading.

How about "in practice, parsers can decide"  instead of "we can decide"?

Reply:
I think this is resolved based on the offline discussion.